samples/guestbook/gcp/deploy/main.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The deploy program builds the Guestbook server locally and deploys it to
+// GKE.
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+func main() {
+	guestbookDir := flag.String("guestbook_dir", "..", "directory containing the guestbook example")
+	tfStatePath := flag.String("tfstate", "terraform.tfstate", "path to terraform state file")
+	flag.Parse()
+	log.SetPrefix("gcp/deploy: ")
+	if err := deploy(*guestbookDir, *tfStatePath); err != nil {
+		fmt.Fprintln(os.Stderr, "deploy:", err)
+		os.Exit(1)
+	}
+}
+
+func deploy(guestbookDir, tfStatePath string) error {
+	type tfItem struct {
+		Sensitive bool
+		Type      string
+		Value     string
+	}
+	type state struct {
+		Project          tfItem
+		ClusterName      tfItem `json:"cluster_name"`
+		ClusterZone      tfItem `json:"cluster_zone"`
+		Bucket           tfItem
+		DatabaseInstance tfItem `json:"database_instance"`
+		DatabaseRegion   tfItem `json:"database_region"`
+		MotdVarConfig    tfItem `json:"motd_var_config"`
+		MotdVarName      tfItem `json:"motd_var_name"`
+	}
+	tfStateb, err := runb("terraform", "output", "-state", tfStatePath, "-json")
+	if err != nil {
+		return err
+	}
+	var tfState state
+	if err := json.Unmarshal(tfStateb, &tfState); err != nil {
+		return fmt.Errorf("parsing terraform state JSON: %v", err)
+	}
+	zone := tfState.ClusterZone.Value
+	if zone == "" {
+		return fmt.Errorf("empty or missing cluster_zone in %s", tfStatePath)
+	}
+	gcp := gcloud{project: tfState.Project.Value}
```

Comment:
This should be moved down, closer to its use.

Reply:
Done.